csharp/src/Ice/OutgoingRequestFrame.cs

```diff
@@ -0,0 +1,149 @@
+            frame.StartEncapsulation(proxy.EncodingVersion, format);
+            return frame;
+        }
+
+        /// <summary>Creates a new outgoing request frame with for in-parameters payload the in-parameters of an
+        /// incoming request frame.</summary>
+        /// <param name="proxy">A proxy to the target Ice object. This method uses the communicator, identity, facet
+        /// and context of this proxy to create the request frame.</param>
+        /// <param name="operation">The operation to invoke on the target Ice object.</param>
+        /// <param name="idempotent">True when operation is idempotent, otherwise false.</param>
+        /// <param name="payload">The payload of this request frame, which represents the marshaled in-parameters.
+        /// </param>
+        /// <param name="context">An optional explicit context. When non null, it overrides both the context of the
+        /// proxy and the communicator's current context (if any).</param>
+        /// <returns>A new OutgoingRequestFrame.</returns>
+        public static OutgoingRequestFrame Create(IObjectPrx proxy, string operation, bool idempotent,
+            ArraySegment<byte> payload, Context? context = null)
+        {
+            var frame = new OutgoingRequestFrame(proxy.Communicator, proxy.Identity, proxy.Facet, operation, idempotent,
+                proxy.Context, context);
+
+            if (payload.Count == 0)
+            {
+                frame.WriteEmptyEncapsulation(proxy.EncodingVersion);
+            }
+            else
+            {
+                // TODO: works only because we know how payload is created!
+                frame.WriteBlob(payload.Array);
+            }
+            frame.IsSealed = true;
+            return frame;
+        }
+
+         /// <summary>Marks the end of the in-parameters for this request frame.</summary>
+        public void EndPayload()
+        {
+            Debug.Assert(!IsSealed); // TODO: throw an exception
+            EndEncapsulation();
+            IsSealed = true;
+            // Note: with the 2.0 encoding the frame will be sealed later, after the context is written.
+        }
+
+        private OutgoingRequestFrame(Communicator communicator, Identity identity, string facet, string operation,
+            bool idempotent, Context? prxContext, Context? context)
+            : base(communicator)
+        {
+            Identity = identity;
+            Facet = facet;
+            Operation = operation;
+            IsIdempotent = idempotent;
+
+            WriteBlob(Protocol.requestHdr);
+            identity.IceWrite(this);
+            if (facet.Length == 0)
+            {
+                WriteStringSeq(System.Array.Empty<string>());
+            }
+            else
+            {
+                string[] facetPath = { facet };
+                WriteStringSeq(facetPath);
+            }
+
+            WriteString(operation);
+            WriteByte(idempotent ? (byte)0x2 : (byte)0x0);
```

Comment:
would be nice to use a constant/or enum here instead of `0x2` and `0x0`

Reply:
That's from a previous PR. Maybe reintroduce an internal OperationMode?

Reply:
Fixed.